File: generated_code/GeminiPro/Python/094_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Python class that implements a dictionary-like set for
    integers using a Splay Tree.

    The main public methods are:
    - insert(key): Adds a key to the tree.
    - delete(key): Removes a key from the tree.
    - search(key): Searches for a key and splays the accessed node.
    """

    class _Node:
        """A private inner class representing a node in the Splay Tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            p = x.parent
            g = p.parent

            if not g:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig (left-left)
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig (right-right)
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag (left-right)
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag (right-left)
                self._right_rotate(p)
                self._left_rotate(g)

    def search(self, key):
        """
        Searches for a key in the tree.

        This method performs the splaying operation on the accessed node (if the
        key is found) or its parent (if the key is not found) to move it to
        the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        node = self.root
        last_node = self.root
        found = False
        
        while node:
            last_node = node
            if key == node.key:
                found = True
                break
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        # Splay the last accessed node (or the found node) to the root
        self._splay(last_node)
        
        return found

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the node with that key is splayed to the
        root. If the key is new, it is inserted as in a standard BST, and
        the new node is then splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        # Case 1: The tree is empty
        if not self.root:
            self.root = self._Node(key)
            return

        # Traverse the tree to find the insertion point or an existing key
        current = self.root
        parent = None
        while current:
            parent = current
            if key == current.key:
                # Key already exists, splay it and we are done
                self._splay(current)
                return
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Case 2: Key not found, insert a new node
        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        If the key is not found, the tree is splayed based on the last
        accessed node during the search, and no deletion occurs. If the key
        is found, it is removed, and the tree is restructured.

        Args:
            key (int): The integer key to delete.
        """
        # Search for the key. If found, it will be splayed to the root.
        if not self.search(key):
            # Key is not in the tree, nothing to delete.
            # search() has already splayed the closest node.
            return

        # At this point, the node to delete is the root due to search()
        node_to_delete = self.root
        
        left_subtree = node_to_delete.left
        right_subtree = node_to_delete.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect the left subtree from the old root
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max element to the root of the left subtree.
            # We temporarily set self.root to the left subtree's root to
            # perform the splay operation within that subtree.
            self.root = left_subtree
            self._splay(max_node)
            
            # After splaying, self.root is the new root of the combined tree.
            # It has no right child, so we attach the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/094_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Inner class to represent a node in the Splay Tree.
     * It includes a parent reference, which is crucial for the splay operation.
     */
    private class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Public API ---

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree is splayed on that key.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node current = root;
        Node parent = null;
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay the node and return
                splay(current);
                return;
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }

        splay(newNode);
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not found, the tree structure is modified by splaying the last accessed node.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Search for the key and splay the last accessed node to the root.
        search(key);

        // After search, if the root's key doesn't match, the key was not in the tree.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // No left child, the right subtree becomes the new tree.
            this.root = rightSubtree;
            if (this.root != null) {
                this.root.parent = null;
            }
        } else {
            // Disconnect the left subtree to work on it.
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree.
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }

            // To use our splay method, we temporarily set the main root to the left subtree's root.
            this.root = leftSubtree;
            splay(maxNode); // maxNode is now the root of the modified left subtree.

            // `this.root` now points to maxNode. Attach the original right subtree.
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }

    /**
     * Searches for a key in the splay tree.
     * This operation performs a splay. If the key is found, the node containing the key
     * is moved to the root. If the key is not found, the last non-null node accessed
     * during the search is moved to the root.
     *
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }

        Node current = root;
        Node lastNode = root; // To splay the last visited node

        while (current != null) {
            lastNode = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key found
                splay(current);
                return key;
            }
        }

        // Key not found, splay the last visited node
        splay(lastNode);

        // After splaying, if the new root's key is not what we looked for,
        // it means the key was not in the tree.
        if (root != null && root.key == key) {
            return key;
        } else {
            return null;
        }
    }

    // --- Private Helper Methods ---

    /**
     * Performs the splay operation on a node x, moving it to the root.
     *
     * @param x The node to splay.
     */
    private void splay(Node x) {
        if (x == null) return;

        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;

            if (grandParent == null) { // Zig case
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else { // Zig-Zig or Zig-Zag
                if (x == parent.left && parent == grandParent.left) { // Zig-Zig (left-left)
                    rightRotate(grandParent);
                    rightRotate(parent);
                } else if (x == parent.right && parent == grandParent.right) { // Zig-Zig (right-right)
                    leftRotate(grandParent);
                    leftRotate(parent);
                } else if (x == parent.right && parent == grandParent.left) { // Zig-Zag (left-right)
                    leftRotate(parent);
                    rightRotate(grandParent);
                } else { // Zig-Zag (right-left)
                    rightRotate(parent);
                    leftRotate(grandParent);
                }
            }
        }
        this.root = x;
    }

    /**
     * Performs a left rotation on the given node.
     *
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node.
     *
     * @param x The node to rotate.
     */
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y != null) {
            x.left = y.right;
            if (y.right != null) {
                y.right.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.right = x;
            x.parent = y;
        }
    }
}